src/transport/io.rs

```diff
@@ -186,3 +186,101 @@ mod tests {
         io.write_buf(&mut "hello".into_buf()).expect("write_buf");
     }
 }
+
+#[derive(Debug)]
+pub struct KafkaIo<I>
+where
+    I: AsyncRead + AsyncWrite,
+{
+    inner: I,
+    start_pos_next_message: usize,
+}
+
+impl<I> KafkaIo<I>
+where
+    I: AsyncRead + AsyncWrite,
+{
+    pub fn new(io: I) -> Self {
+        KafkaIo {
+            inner: io,
+            start_pos_next_message: 0,
+        }
+    }
+}
+
+fn analyze_kafka_message(buf: &[u8]) -> i32 {
+    let decode_result = kafka_codec::primitives::DecodedRequest::decode(buf);
+    match decode_result {
+        Ok(decoded_reuquest) => {
+            println!("Message decoded successfully;");
+            println!("Message Size: {}", decoded_reuquest.size);
+            println!("Message Header: {}", decoded_reuquest.header);
+            println!("Message Body: {:?}", decoded_reuquest.body);
+            return decoded_reuquest.size;
+        }
+        Err(e) => {
+            println!("Error when decoding: {:?} ; buf: {:?}", e, buf);
+            return -1;
+        }
+    }
+}
+
+impl<I> io::Read for KafkaIo<I>
+where
+    I: AsyncRead + AsyncWrite,
+{
+    // Is it possible to read by frame? If so, then we need to handle the frame case
+    // for example, a Kafka message is splitted into two frames
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let result = self.inner.read(buf);
+        match result {
+            Ok(n) => {
+                while self.start_pos_next_message <= n {
```

Comment:
`start_pos_next_message` is the state of the position of the next message to analyze.